File: backend/app/api/routes/bins.py
import uuid
from typing import Any

from fastapi import APIRouter, Depends, HTTPException
from app import crud
from app.api.deps import SessionDep, get_current_active_superuser
from app.models import BinCreate, BinPublic, BinUpdate, Message

router = APIRouter(prefix="/bins", tags=["bin"])


@router.get(
    "/",
    dependencies=[Depends(get_current_active_superuser)],
    response_model=list[BinPublic],
)
def read_bins(session: SessionDep, skip: int = 0, limit: int = 100) -> Any:
    """
    Retrieve bins.
    """
    bins = crud.get_bins(session=session, skip=skip, limit=limit)
    return bins


@router.post(
    "/", dependencies=[Depends(get_current_active_superuser)], response_model=BinPublic
)
def create_bin(*, session: SessionDep, bin_in: BinCreate) -> Any:
    """
    Create new bin.
    """
    bin_obj = crud.create_bin(session=session, bin_create=bin_in)
    return bin_obj


@router.get(
    "/{bin_id}",
    dependencies=[Depends(get_current_active_superuser)],
    response_model=BinPublic,
)
def read_bin_by_id(bin_id: uuid.UUID, session: SessionDep) -> Any:
    """
    Get a specific bin by id.
    """
    bin_obj = crud.get_bin_by_id(session=session, bin_id=bin_id)
    if not bin_obj:
        raise HTTPException(status_code=404, detail="Bin not found")
    return bin_obj


@router.patch(
    "/{bin_id}",
    dependencies=[Depends(get_current_active_superuser)],
    response_model=BinPublic,
)
def update_bin(
    *, session: SessionDep, bin_id: uuid.UUID, bin_in: BinUpdate
) -> Any:
    """
    Update a bin.
    """
    bin_obj = crud.get_bin_by_id(session=session, bin_id=bin_id)
    if not bin_obj:
        raise HTTPException(status_code=404, detail="Bin not found")
    update_data = bin_in.model_dump(exclude_unset=True)
    bin_obj.sqlmodel_update(update_data)
    session.add(bin_obj)
    session.commit()
    session.refresh(bin_obj)
    return bin_obj


@router.delete(
    "/{bin_id}",
    dependencies=[Depends(get_current_active_superuser)],
    response_model=Message,
)
def delete_bin(session: SessionDep, bin_id: uuid.UUID) -> Any:
    """
    Delete a bin.
    """
    bin_obj = crud.get_bin_by_id(session=session, bin_id=bin_id)
    if not bin_obj:
        raise HTTPException(status_code=404, detail="Bin not found")
    session.delete(bin_obj)
    session.commit()
    return Message(message="Bin deleted successfully")


@router.get(
    "/warehouse/{warehouse_id}",
    dependencies=[Depends(get_current_active_superuser)],
    response_model=list[BinPublic],
)
def read_bins_by_warehouse(warehouse_id: uuid.UUID, session: SessionDep, skip: int = 0, limit: int = 100) -> Any:
    """
    Get bins for a specific warehouse.
    """
    bins = crud.get_bins_by_warehouse(session=session, warehouse_id=warehouse_id, skip=skip, limit=limit)
    return bins

File: frontend/src/lib/bins.ts
import axios from "axios"
import { OpenAPI } from "@/client"

const apiBase = OpenAPI.BASE.replace(/\/$/, "")

const getAuthHeaders = () => {
  const token = localStorage.getItem("access_token")
  return token ? { Authorization: `Bearer ${token}` } : {}
}

export type BinPublic = {
  id: string
  nombre: string
  x: string
  y: string
  z: string
  warehouse_id: string
  is_active: boolean
  created_at?: string | null
  updated_at?: string | null
}

export type BinCreate = {
  nombre: string
  x: string
  y: string
  z: string
  warehouse_id: string
  is_active?: boolean
}

export type BinUpdate = {
  nombre?: string
  x?: string
  y?: string
  z?: string
  warehouse_id?: string
  is_active?: boolean
}

export const BinsService = {
  readBins: async ({ skip = 0 } = {}) => {
    const response = await axios.get<BinPublic[]>(`${apiBase}/api/v1/bins`, {
      headers: getAuthHeaders(),
      params: { skip },
    })
    return response.data
  },

  createBin: async (data: BinCreate) => {
    const response = await axios.post<BinPublic>(`${apiBase}/api/v1/bins`, data, {
      headers: getAuthHeaders(),
    })
    return response.data
  },

  updateBin: async (id: string, data: BinUpdate) => {
    const response = await axios.patch<BinPublic>(`${apiBase}/api/v1/bins/${id}`, data, {
      headers: getAuthHeaders(),
    })
    return response.data
  },

  deleteBin: async (id: string) => {
    const response = await axios.delete(`${apiBase}/api/v1/bins/${id}`, {
      headers: getAuthHeaders(),
    })
    return response.data
  },

  readBinsByWarehouse: async (warehouseId: string, { skip = 0 } = {}) => {
    const response = await axios.get<BinPublic[]>(`${apiBase}/api/v1/bins/warehouse/${warehouseId}`, {
      headers: getAuthHeaders(),
      params: { skip },
    })
    return response.data
  },
}